src/travel-details-screens/TripDetailsScreenComponent.tsx

```diff
@@ -151,73 +151,155 @@ export const TripDetailsScreenComponent = ({
 
   if (
-    !(
-      fromTripsSearchToTicketEnabled &&
-      fromTariffZoneWeSellSingleTicketsFor &&
-      toTariffZoneWeSellTicketFor
-    ) ||
+    !enable_ticketing ||
+    !isFromTravelSearchToTicketEnabled ||
     hasTooLongWaitTime
   )
     return;
 
-  const tariffZoneTo: TariffZoneWithMetadata = {
-    resultType: 'zone',
-    venueName: nonHumanLegs[nonHumanLegs.length - 1]?.toPlace?.name,
-    ...toTariffZoneWeSellTicketFor,
-  };
-  const tariffZoneFrom: TariffZoneWithMetadata = {
-    resultType: 'zone',
-    venueName: nonHumanLegs[0]?.fromPlace?.name,
-    ...fromTariffZoneWeSellSingleTicketsFor,
-  };
+  const nonHumanLegs = getNonHumanLegs(tripPattern.legs);
+  const ticketStartTime = calculateTicketStartTime(nonHumanLegs);
+
+  const ticketInfoForBus = getTicketInfoForBus(
+    tripPattern,
+    nonHumanLegs,
+    fareProductTypeConfigs,
+    tariffZones,
+    ticketStartTime,
+  );
+  if (ticketInfoForBus) return ticketInfoForBus;
+
+  const ticketInfoForBoat = getTicketInfoForBoat(
+    tripPattern,
+    nonHumanLegs,
+    fareProductTypeConfigs,
+    harbors,
+    ticketStartTime,
+  );
+  if (ticketInfoForBoat) return ticketInfoForBoat;
+}
+
+function getNonHumanLegs(legs: Leg[]) {
+  return legs.filter(
+    (leg) => leg.mode !== Mode.Foot && leg.mode !== Mode.Bicycle,
+  );
+}
+
+function calculateTicketStartTime(legs: Leg[]) {
+  const tripStartWithBuffer = addMinutes(parseISO(legs[0]?.aimedStartTime), -5);
+  return tripStartWithBuffer.getTime() <= Date.now()
+    ? undefined
+    : formatISO(tripStartWithBuffer);
+}
 
-  // modes we can sell single tickets for. Might not always match modes we sell tickets for,
-  // as from travel search to ticket currently only supports single ticket
-  const someLegsAreNotSingleTicket = hasLegsWeCantSellTicketsFor(tripPattern, [
+function getTicketInfoForBus(
+  tripPattern: TripPattern,
+  nonHumanLegs: Leg[],
+  fareProductTypeConfigs: FareProductTypeConfig[],
+  tariffZones: TariffZone[],
+  ticketStartTime?: string,
+) {
+  const canSellCollab = canSellCollabTicket(tripPattern);
+  const hasOnlyValidBusLegs = !hasLegsWeCantSellTicketsFor(tripPattern, [
     'cityTram',
     'expressBus',
     'localBus',
     'localTram',
     'regionalBus',
     'shuttleBus',
   ]);
-  if (!enable_ticketing || (someLegsAreNotSingleTicket && !canSellCollab))
-    return;
 
-  const tripStartWithBuffer = addMinutes(
-    parseISO(nonHumanLegs[0]?.aimedStartTime),
-    -5,
+  if (!hasOnlyValidBusLegs && !canSellCollab) return;
+
+  const fromTariffZone = getTariffZoneWithMetadata(
+    nonHumanLegs[0].fromPlace,
+    tariffZones,
   );
-  const ticketStartTime =
-    tripStartWithBuffer.getTime() <= Date.now()
-      ? undefined
-      : formatISO(tripStartWithBuffer);
+  const toTariffZone = getTariffZoneWithMetadata(
+    nonHumanLegs[nonHumanLegs.length - 1].toPlace,
+    tariffZones,
+  );
+
+  if (!fromTariffZone || !fromTariffZone) return;
```

Comment:
One of these should probably be `toTariffZone`, I would guess.

```suggestion
  if (!fromTariffZone || !toTariffZone) return;
```